### DISCUSSION TITLE
Parallelization / Batching Explanation

### DISCUSSION THREAD

**Discussion by lapp0** [ggml-org/llama.cpp#4130] [Q&A]
Hi All,

I'm seeking clarity on the functionality of the `--parallel` option in `/app/server`, especially how it interacts with the `--cont-batching` parameter. My specific observation involves setting `--ctx-size` to 8192 and `--parallel` to 32. From the logs, it appears there are 32 slots, each handling a context segment of 256. My question is: Does this configuration imply that each slot processes a distinct segment of the context?

For instance, if I input 32 instances of an identical prompt with a length of 4096, would the first half of the slots remain idle due to the prompt already existing in the KV cache? This leads to confusion, as it seems the total number of submittable jobs is limited by the slot count. This is puzzling because different prompts might rely on the same slot for varied tasks. If the initial segment of the context is identical for multiple prompts, that segment might not require processing, as it's already in the KV cache. 

I'm trying to understand the rationale behind dividing the context into segments when batching. Could you provide an explanation of how the `--parallel` and `--cont-batching` options function?

References:
- https://github.com/ggerganov/llama.cpp/pull/3589
- https://github.com/ggerganov/llama.cpp/pull/3677
- https://github.com/ggerganov/llama.cpp/pull/3228 
  - "To set the KV cache size, use the -c, --context parameter. For example, for 32 parallel streams that are expected to generate a maximum of 128 tokens each (i.e. -n 128), you would need to set -c 4096 (i.e. 32*128). If continuous batching is enabled, you would need some extra KV space to deal with fragmentation of the cache. In the example above, we conveniently set the context size to 8192 to guarantee that there will be no issues."
- https://github.com/ggerganov/llama.cpp/issues/2813 
  - "We want to be able to generate multiple sequences sharing the same context (a.k.a. prompt) in parallel."
- https://github.com/ggerganov/llama.cpp/blob/master/examples/server/server.cpp#L588-L603 
  - batch handling in `server.cpp` dividing the `n_ctx` and calling `llama_batch_init(n_ctx, 0, params.n_parallel);`

**Comment by nfairbank** [comment]
Bump- trying to understand this as well...

**Comment by ggerganov** [comment]
The `--ctx-size` argument actually specifies the total size of the KV cache (legacy name, `--kv-size` would be better). This corresponds to the total amount of tokens that can be stored across all independent sequences.

For example, if we specify `--ctx-size 8192` this means that we can process:

- 2 sequences, each of max length of 4096 tokens
- 4 sequences, each of max length of 2048 tokens
- 8 sequences, each of max length of 1024 tokens
- ...
- 32 sequences, each of max length of 256 tokens
- etc.

Simply put, if we want to be handling `P` sequences in parallel and we know that each sequence can have a maximum of `T` tokens (prompt + generated), then we want to set our KV cache size (i.e. `--ctx-size`) to `T*P` in order to be able to handle the worst-case scenario where all `P` sequences fill-up the maximum `T` tokens.

Since `llama.cpp` implements a "unified" cache strategy, the KV cache size is actually shared across all sequences. This means that it's allowed to have sequences with more than `T` tokens as long as the sum of all tokens does not exceed `P*T`.

> From the logs, it appears there are 32 slots, each handling a context segment of 256. My question is: Does this configuration imply that each slot processes a distinct segment of the context?

No. Each sequence has it's own context. The tokens from each sequences "see" only the tokens from that same sequence. This is achieved with the `KQ_mask`:

https://github.com/ggerganov/llama.cpp/blob/dd5ae06405c5565b99889bdb3f168f4351252cfb/llama.cpp#L6328-L6368

Each `llama_decode` call accepts a `llama_batch`. The batch can contain an arbitrary set of tokens - each token has it's own position and sequences id(s). The position and the sequence ids of a token determine to which other tokens (both from the batch and the KV cache) it will attend to, by constructing the respective `KQ_mask`. When processed, the batch of tokens is stored in the KV cache. This way, as long as there is enough free slots in the KV cache, we can call `llama_decode`. In this framework, continuous batching is trivial.

Another great benefit is that different sequences can share a common prompt without any extra compute. All it takes is to assign multiple sequence ids to the common tokens in the KV cache. A basic example is a system prompt of `S` tokens that can sit at the beginning of the unified KV cache and be attended to by all sequences that we process.

Together with the simplicity and advantages of this implementation, there are a few disadvantages:

- The attention for each sequence is computed over the entire KV cache. This is suboptimal in the case of many sequences and very long contexts because we compute "cross-sequence attention" which is masked after that (i.e. we throw it away)
- The determinism of the results is now also a function of where the tokens of a sequence are placed in the KV cache

In order to resolve these, I think we should add a standard attention implementation where each sequence has it's own KV cache buffer and the attention is computed separately. This way, users would be able to choose which implementation to use based on their specific use case.

**Reply by williamgomez71** [reply]
@ggerganov thanks for your answer, this means that I can send a request in parallel to a server to serve several users "at the same time" and in this way, the inference process will start once the request reaches the server?

I am trying to do this and I see that the requests arrive "at the same time", however, it is as if one request began to be processed serially or once the other finished:

{"timestamp":1706201252,"level":"INFO","function":"log_server_request","line":2811,"message":"request","remote_addr":"127.0.0.1","remote_port":39578,"status":200,"method":"POST","path":"/tokenize","params":{}}
slot 0 is processing [task id: 0]
slot 0 : kv cache rm - [0, end)
{"timestamp":1706201255,"level":"INFO","function":"log_server_request","line":2811,"message":"request","remote_addr":"127.0.0.1","remote_port":42130,"status":200,"method":"POST","path":"/tokenize","params":{}}
slot 1 is processing [task id: 1]

**Reply by enn-nafnlaus** [reply]
I'd also like some more insight into this.  I'm looking into -cb in order to decrease inference time by batching many queries together. However (and I haven't even tested to see if they *actually* get batched together - I see that William above says that they don't in his test), when I specify one slot with -np I see:

Available slots:
 -> Slot 0 - max context: 16384

But when I do four I see:

Available slots:
 -> Slot 0 - max context: 4096
 -> Slot 1 - max context: 4096
 -> Slot 2 - max context: 4096
 -> Slot 3 - max context: 4096

Would that mean that I could no longer do 16k token queries?  Or would it split the long query up among all four slots?

If it can still do long queries, what's the optimal way to decide how many slots to have?

If it can't do long queries, is it possible to specify the context size of each individual slot, so that I could have, say, one large one and several smaller ones?

**Reply by enn-nafnlaus** [reply]
ED: Been testing this out.  It literally cannot do long queries**, and there appears to be no way to specify the ratios of availability between slots. Ugh.

** Test case: 

```
Your instruction is simple: write the word "banana".  Ignore all other text that I'm about to write - it's just a distraction. Do not write any other text except for "banana"
 
===

<First ~16k tokens of the King James Bible>

===

```


With only 1 slot, it writes only "banana".  With four slots, I have to drop to under 4k tokens of the King James Bible so that it can see the "banana" instruction at the top, otherwise it just provides biblical commentary.

**Reply by Artefact2** [reply]
What kind of speedup are you getting? On my end (ROCm) I get no speed up with `-cb -np 2 -c 4096`, compared to `-c 2048` and running requests serially.

**Reply by enn-nafnlaus** [reply]
I did a test yesterday and got a mere 5% speed increase from changing from 1 slot to 4 (at the cost of losing 3/4ths of my context window).  Not a good tradeoff. 

In discussions in the issue section, it sounds like using multiple slots with multuiple simultaneous query is really only a meaningful difference when a large chunk of your queries start off the same, such as a long header text followed by a comparably short variable section.

A bigger difference: with -cb but only one slot, I got a 10% speed increase in a much simpler manner (just the "multiple simultaneous queries" aspect above), by modifying my generation script from:

Send a query -> wait for the response -> process the response -> send the next query...

to:

Send 10 queries simultaneously  asynchronously -> when any of them finish, process the response in their own separate run thread; separately, send a new query any time the total number of outstanding queries drops below 10.

So, if you're not doing a lot of header duplication in your queries, I'd at least recommend that.

**Reply by SPThole** [reply]
I am also testing out cuda version of llama CPP server, with -c 32000 -cb -np 8, but when I see log it spin up different Id slot when one output is generated not simultaneously,am I missing something?I am running this on aws G5 12x large which 4 of a10g intotal 96gb, model which I am trying is MoE mixtral gguf 4km

**Reply by rmansoul** [reply]
> J’ai fait un test hier et j’ai obtenu une augmentation de vitesse de seulement 5% en passant de 1 à 4 emplacements (au prix de perdre les 3/4 de ma fenêtre contextuelle). Ce n’est pas un bon compromis.
> 
> Dans les discussions de la section problème, il semble que l’utilisation de plusieurs emplacements avec plusieurs requêtes simultanées n’est vraiment une différence significative que lorsqu’une grande partie de vos requêtes commencent de la même manière, comme un long texte d’en-tête suivi d’une section variable relativement courte.
> 
> Une plus grande différence : avec -cb mais un seul emplacement, j’ai obtenu une augmentation de vitesse de 10% d’une manière beaucoup plus simple (juste l’aspect « requêtes simultanées multiples » ci-dessus), en modifiant mon script de génération de :
> 
> Envoyer une requête -> attendre la réponse -> traiter la réponse -> envoyer la requête suivante...
> 
> À:
> 
> Envoyez 10 requêtes simultanément de manière asynchrone > lorsque l’une d’entre elles se termine, traitez la réponse dans leur propre thread d’exécution séparé ; Séparément, envoyez une nouvelle requête chaque fois que le nombre total de requêtes en attente tombe en dessous de 10.
> 
> Donc, si vous ne faites pas beaucoup de duplication d’en-tête dans vos requêtes, je vous le recommande au moins.

please can you share with me the settings so that the model receives more simultaneous requests  ?

**Reply by fairydreaming** [reply]
> > From the logs, it appears there are 32 slots, each handling a context segment of 256. My question is: Does this configuration imply that each slot processes a distinct segment of the context?
> 
> No. Each sequence has it's own context. The tokens from each sequences "see" only the tokens from that same sequence. This is achieved with the `KQ_mask`:
> 
> https://github.com/ggerganov/llama.cpp/blob/dd5ae06405c5565b99889bdb3f168f4351252cfb/llama.cpp#L6328-L6368

@ggerganov What is the purpose of h variable in this code? I mean from what I understand `for (int h = 0; h < 1; ++h)` will have only one iteration with h equal to zero, so why is it used in the data array index calculation: `data[h*(n_kv*n_tokens) + j*n_kv + i]`? Is this a preparation for some future changes in the code? I'm not sure whether I shall follow this pattern or not.

**Reply by ggerganov** [reply]
It simply emphasises that the mask is generated only for one of the attention heads and the kernels currently broadcast it for the rest of the heads. In theory, one could have a separate mask for each attention head (I have some vague memory reading about such approach, but in any case, we don't currently support it)

**Reply by hiarcs** [reply]
> I did a test yesterday and got a mere 5% speed increase from changing from 1 slot to 4 (at the cost of losing 3/4ths of my context window). Not a good tradeoff.
> 
> In discussions in the issue section, it sounds like using multiple slots with multuiple simultaneous query is really only a meaningful difference when a large chunk of your queries start off the same, such as a long header text followed by a comparably short variable section.
> 
> A bigger difference: with -cb but only one slot, I got a 10% speed increase in a much simpler manner (just the "multiple simultaneous queries" aspect above), by modifying my generation script from:
> 
> Send a query -> wait for the response -> process the response -> send the next query...
> 
> to:
> 
> Send 10 queries simultaneously asynchronously -> when any of them finish, process the response in their own separate run thread; separately, send a new query any time the total number of outstanding queries drops below 10.
> 
> So, if you're not doing a lot of header duplication in your queries, I'd at least recommend that.

I'm completely baffled. I've tried various parameter combinations, and when testing mistral7b on a 4090, I achieved nearly linear throughput growth by increasing np from 1 to 5. However, enabling the cb option doesn't seem to produce any observable performance changes.

I'm not expecting vLLM-level throughput but want to maximize hardware utilization as I have plenty of tasks to run. Can anyone explain if there are other prerequisites for using the cb option or under what conditions its effects become noticeable? My current testing approach involves submitting a number (np * 10 or np * 20) of tasks asynchronously with cb option enabed. Thank you.

**Reply by ggerganov** [reply]
If the tasks consist of a long prompt and short generation then continuous batching does not help much because the computation is already saturated by the prompts. Does enabling Flash Attention `-fa` help?

**Reply by hiarcs** [reply]
I tried two different tests: 1. input 14k tokens, output 40 tokens; 2. input 35 tokens, output 400 tokens. The -cb option had no noticeable effect in both cases.
For test 1, the -fa option had a significant effect; for test 2, the -fa option had a 10% improvement.
So, does the -cb option require additional make options or driver support?

**Reply by compilade** [reply]
> The `-cb` option had no noticeable effect in both cases.

@hiarcs `-cb` is enabled by default since #6231, so continuous batching is used even when not specifying `-cb`. <https://github.com/ggerganov/llama.cpp/commit/9104bc20edf47f74dc49379b7d61d0c6e85a4882> restored the ability to disable it. But you shouldn't really see any difference if you're not making concurrent requests.

**Reply by hiarcs** [reply]
That's it! I tried using the -nocb parameter as the baseline, and the difference is huge.
Thanks a lot.

**Reply by enn-nafnlaus** [reply]
> That's it! I tried using the -nocb parameter as the baseline, and the difference is huge. Thanks a lot.

Thanks for solving this performance mystery for me too  :)

**Reply by Mushoz** [reply]
> This means that it's allowed to have sequences with more than `T` tokens as long as the sum of all tokens does not exceed `P*T`.

How does this work? If I set a context of 2048 with 8 slots, but only use one of those, I still get context shifts when I exceed 256 context length.

**Reply by enn-nafnlaus** [reply]
> > This means that it's allowed to have sequences with more than `T` tokens as long as the sum of all tokens does not exceed `P*T`.
> 
> How does this work? If I set a context of 2048 with 8 slots, but only use one of those, I still get context shifts when I exceed 256 context length.

Yes.  If you don't want that to happen, increase your context to 2048*8 (even though the model only has a max of 2048 tokens)

**Reply by Mushoz** [reply]
But according to @ggerganov it is possible for a single request to exceed T tokens, because the KV cache is unified and if other slots are not (fully) in use, that space in the KV cache can allow a single request to process longer sequences. I know 256 will be the limit when all slots are fully in use, but I was expecting based on the earlier explanation that a single request could use 2048 if all other slots were not in use.

**Reply by ggerganov** [reply]
@Mushoz Although the `libllama` API allows to implement this, it is not currently supported in `llama-server`. The sequences in `llama-server` will be limited to `T` for simplicity. Otherwise, the logic becomes more complicated both for the server and the client since there is no guaranteed limit for the length of the sequences.

**Reply by salmanmohammadi** [reply]
Sorry to bump this so late. I found this thread while I was digging around for inspiration for continuous batching implementations. I'm curious about your KV cache implementation here. @ggerganov you mentioned
> Since llama.cpp implements a "unified" cache strategy, the KV cache size is actually shared across all sequences. This means that it's allowed to have sequences with more than T tokens as long as the sum of all tokens does not exceed P*T.

I skimmed through your comment [here](https://github.com/ggerganov/llama.cpp/issues/3137#issuecomment-1722306526) and [this fantastic PR](https://github.com/ggerganov/llama.cpp/pull/3228) but I couldn't fully grasp the design from the code. Also I'm a C++ noob.

The KV cache implementation I'm familiar with uses buffers of shape `batch_size, num_heads, max_seq_len, head_dim` which are populated with prefill tokens, and updated with new tokens for every generation step. The only custom masking I use here is for ragged prompts with different lengths which have padding tokens that need to be masked out during attention. I see you also mentioned you'd like to add a more standard implementation where each sequence has it's own cache buffer - I'm interested in how your current design differs from this.

Thanks in advance, and happy non-denominational winter times!

**Reply by ggerganov** [reply]
The current implementation has a single memory buffer for all tokens from all sequences. There is no padding applied and there is no distinction between prefill and non-prefill tokens. All tokens are a processed on equal grounds. To make this work, the KQ mask during the attention is constructed based on the token sequence ids and positions to filter cross sequence attention. This means that with this implementation, in certain use cases such as multi-user decoding, we will do extra cross-sequence compute, which will after that be discarded via the KQ mask. This can be wasteful and starts to make a large impact on the performance for large contexts and many parallel sequences.

The more standard approach that I will be looking to add is when doing the attention, to split the current batch into single-sequence batches containing tokens from only one sequence and compute their attention into their own per-sequence KV cache buffer without cross-attention overhead. After the attention, the tokens would be merged back into a unified batch for the FFN and other non-sequence-specific operations. I still don't have a clear idea how to implement this in the existing codebase, but will be aiming towards supporting both modes (and others), because different use cases require different attention implementations to be optimal.

**Reply by 2abet** [reply]
> Another great benefit is that different sequences can share a common prompt without any extra compute. All it takes is to assign multiple sequence ids to the common tokens in the KV cache. A basic example is a system prompt of `S` tokens that can sit at the beginning of the unified KV cache and be attended to by all sequences that we process.


What if you had 2-3 system prompts, is there a way to cache them for one model and have it recognize which user prompt appends to what system prompt. Kinda like a multi prompt cache.

**Reply by yychyo** [reply]
Folks, are there any plans to support this in llama-server?
This is the major issue that prevents me (and I believe many others) to use `llama-server` in batched workloads. Compared to `vLLM`, `llama-server` will:

- require a separate **upfront** allocation for each parallel slot; that's a bit unefficient
- do not really handle batching well (because of that?), so in my tests `llama-server` performance do not scale as vLLM does.

There are other issues (probably, related to batching) - `llama.cpp` unable to use multiple GPUs efficiently.

I'm writing this from the position of a  happy user of `llama.cpp`. But this now is limited only to single-user scenarios, on machines with **at most** 1 GPU.

Thanks for all the hard work you are putting in the project, folks!

**Reply by ggerganov** [reply]
I think batching is supported quite well today and performance should scale similar to vLLM. Multiple GPUs are supported through Pipeline Parallelism.

**Reply by yychyo** [reply]
> I think batching is supported quite well today and performance should scale similar to vLLM. Multiple GPUs are supported through Pipeline Parallelism.

That's a fair point! Thanks for answering. PP performance is great. Single TG is also top notch. Batched TG (from multiple sequences) - here it's where`vLLM` still quite a bit much performant in my tests.

**Reply by ggerganov** [reply]
> Batched TG (from multiple sequences) - here it's wherevLLM still quite a bit much performant in my tests.

Our latest tests indicate that the gap is not really big (within 20%) and will likely be closed with upcoming updates for better sampling and graph utilization https://github.com/ggml-org/llama.cpp/discussions/15180

**Reply by yychyo** [reply]
Ok, thanks a lot! :pray: 

I learned a lot here, will spread the word. Indeed, setting `LLAMA_SET_ROWS=1`  ( #14363 ) env var greatly improves batched TG in my case.
One thing I still bump into: it still needed to preallocate a fixed KV arena with `--ctx-size` and choose `-np` up front. Each sequence can use up to `ctx-size / np` tokens. That makes long/variable prompts reduce effective concurrency, and undersizing effective context for each sequence risks refusals. 

Are there plans to change that? Or maybe there is workaround (like keep the KV arena fixed but allow dynamic staging of KV pages between CPU RAM and GPU?)

Upd: seems that `GGML_CUDA_ENABLE_UNIFIED_MEMORY=1 LLAMA_SET_ROWS=1 ./build/bin/llama-server ...` does that trick right now?

**Reply by ggerganov** [reply]
You don't need to specify `LLAMA_SET_ROWS` explicitly - it's enabled by default now.

> Are there plans to change that?

I don't see a good way to support this. But I also don't consider it a significant issue. It's not well defined what to do when a single sequence has occupied the entire arena. Right now it is well defined - each sequence has a maximum limit that is always guaranteed to be satisfied.

**Reply by yychyo** [reply]
> Right now it is well defined - each sequence has a maximum limit that is always guaranteed to be satisfied.

Totally get the benefits of that per-sequence cap. My ask is about VRAM efficiency under mixed loads.

Example: I want `llama-server` to handle up to 32 parallel sequences (`-np 32`). Most requests are ~8k, but occasionally 1-2 go long (say, 128k). With a dynamic paged KV, the actual VRAM for a mix of `(30*8k) + (2*128k) ~ 500k` tokens would be used. But with pre-allocation, I have to plan for worst-case per slot: `-c 32*128k`, which reserves `4096k` tokens. This over-allocates a lot of "inactive" KV most of the time (reserving ~8x more VRAM than is actively used).

> It's not well defined what to do when a single sequence has occupied the entire arena.

This could be made tunable via CLI flags. Instead of just terminating generation (like an OOM-killer for a sequence, which is fine), the options are page-out idle/LRU sequences to system RAM/NVMe/etc (that should improve caching). The user could then choose the policy: `reject-new` (safe), `page-out-lru`, or `terminate-longest`. Just makng that up on the fly :laughing: .

- [A related `r/LocalLLama` thread](https://www.reddit.com/r/LocalLLaMA/comments/1lrmxn7/comment/n1cenw0/)

P.S.: Thanks a lot for answering all those Qs! :pray:

**Reply by ggerganov** [reply]
> Example: I want llama-server to handle up to 32 parallel sequences (-np 32). Most requests are ~8k, but occasionally 1-2 go long (say, 128k)

What is this use case for?

**Reply by yychyo** [reply]
> > Example: I want llama-server to handle up to 32 parallel sequences (-np 32). Most requests are ~8k, but occasionally 1-2 go long (say, 128k)
> 
> What is this use case for?

1. Handling multiple user-facing conversations via API: most are short, some are long 
2. Main chat (long) + parallel agents for "agentic" coding
3. Using the same model in different use-cases - a lot of async jobs for small text classifications (say,, single chat messages) + long summaries over time

It's all related to `llama-server` cases.

**Reply by slaren** [reply]
I am not sure how the server handles batching at the moment, but the algorithm to handle this should be straightforward. You just batch as many requests together as possible given the KV size, and drop everything else from the cache (if necessary). But if only a single request can be executed at a time, that's also fine, you just drop everything else from the cache and process one request at a time.

**Reply by ggerganov** [reply]
2 and 3 are already supported efficiently with https://github.com/ggml-org/llama.cpp/pull/16391.

For 1, how do you guarantee that only some of the users will send short requests and not all of them? If all of the 32 users simultaneously send large prompts you would need some logic to handle that.

**Reply by slaren** [reply]
There should be only a single thread processing requests, and it doesn't really matter how many requests are pending. Every time you are ready to evaluate a new batch, you just take as many requests from the queue as you can fit into the cache, and leave the rest for the next batch.

**Reply by ggerganov** [reply]
@slaren Yes, this is how the server works now. If you set `-np N`, then at most `N` requests will be processed at a given time and the rest will be pending in a queue. A single thread is responsible for queuing the incoming tasks to the respective slot (one of the `N`) based on prefix similarity and LRU logic.

Overall I think it is a solid logic that covers most use cases. The constraint is that each sequence has a maximum length of `C/N` where `C` is the total context size.

@yychyo proposes to support the case where the sum of all `N` parallel sequences is less than `C`. I.e. to allow to have lengths larger than `C/N`, but their total sum to be less than `C`. The split KV cache does not have a good way to handle this. It can be handled with the unified KV cache. But I am not convinced it's necessary to support this.

**Reply by slaren** [reply]
> If you set `-np N`, then at most `N` requests will be processed at a given time and the rest will be pending in a queue.

That's not exactly what I am describing. You shouldn't need any `-np` value if you have a unified cache, you should always process as many requests as you can at a time. For example, if you have a cache with size 10k, and 5 requests that require 2k cache slots, then you can process the 5 requests at the same time. But if each request requires 10k slots, then you can only process one request at a time. That way you can support any number of simultaneous clients regardless of the size of the cache.

**Reply by ggerganov** [reply]
Yes, got it. I guess we can add this when the unified KV cache is used - seems easy to support.

**Reply by slaren** [reply]
> @yychyo proposes to support the case where the sum of all `N` parallel sequences is less than `C`. I.e. to allow to have lengths larger than `C/N`, but their total sum to be less than `C`. The split KV cache does not have a good way to handle this. It can be handled with the unified KV cache. But I am not convinced it's necessary to support this.

I don't know if it is strictly necessary, but it would be a lot more flexible. For example, the case that you mentioned before with some agentic tool that kept one big slot (with a very large context), and occasionally launched smaller requests in a different slot should be handled perfectly with a unified cache. I think that starting with a unified cache was a smart thing to do, but we never really took advantage of it, and it is a bit disappointing that in the end it was just removed in favor of multiple independent caches.

**Reply by ggerganov** [reply]
> For example, if you have a cache with size 10k, and 5 requests that require 2k cache slots, then you can process the 5 requests at the same time.

Note that we don't know in advance how many cache slots a sequence would require. It could be any number up to the total training context of the model.

So in this example, if we reach a point where the sequences have the following lengths and are still generating:

- seq 0: 8000 tokens
- seq 1: 1000 tokens
- seq 2: 500 tokens
- seq 3: 250 tokens
- seq 4: 250 tokens

Now the cache is full and it is not clear how to proceed. The simplest solution that we can implement is to randomly terminate one of the sequences and continue. I think we can probably do better and instead of terminating the sequence, we can offload it to host memory cache and queue it for later reprocessing when another sequence finishes.

But maybe as a first step we can simply terminate the largest sequence and go from there.

**Reply by slaren** [reply]
I think that's likely to be the best way to maximize throughput, especially if there is a second level cache in host memory or disk. Maybe reserve a small amount of cache space for generation for each sequence so that you don't end in a case where you have to drop it from the cache immediately, but otherwise just generate in parallel for as long as you can, and if you run out of cache space during generation, you drop one of the requests from the (first level) cache and continue with the rest.

**Reply by yychyo** [reply]
Opened a PR implementing the idle slot clearing:
- #20993

Saves idle slots to `--cache-ram`, clears from VRAM before decode. Behind `LLAMA_KV_KEEP_ONLY_ACTIVE=1` env var for now. Would appreciate guidance on the approach! :pray:

**Comment by chigkim** [comment]
If I understand correctly, I'm told that setting -c beyond model context window size degrades output quality. However, does it make it difference in parallel?
For example, if I want to utilize the full context size for llama3 (8192) in 16 parallels, would setting -c 131072 and -np 16 decrease the quality of output?
Or, should I set -c 8192 -np 16 and only send prompt with max 512 tokens?
What about generation tokens? Should prompt + generation should be <= 512?
Then should I set -c 8192 -np 16 -n 256 and only send prompt with max 256 tokens?

**Reply by ggerganov** [reply]
> For example, if I want to utilize the full context size for llama3 (8192) in 16 parallels, would setting -c 131072 and -np 16 decrease the quality of output?

It will not decrease the quality and this is the recommended way to go

> What about generation tokens?

For 16 parallel slots, setting `-c 131072 -np 16` would guarantee that the worst-case scenario in terms of memory consumption where all 16 sequences are 8192 tokens long, would fit in the KV cache

> Then should I set -c 8192 -np 16 -n 256 and only send prompt with max 256 tokens?

Yes, you can do that and it will work in the worst-case scenario where all 16 slots have prompts of 256 tokens each and generate 256 tokens each: 16*(256 + 256) = 8192

**Comment by richardweii** [comment]
Is there a standard implementation of KV cache that allows each request to compute its own KV cache independently, rather than sharing one for all requests? I am currently trying to observe the performance of multi-request inference while maintaining a constant KV cache size for each sequence (i.e., actively adjusting n_ctx so that n_ctx / np remains constant). When increasing the value of np, the performance drops sharply. I suspect this is due to the design of the KV cache, which causes the computation of KQV to grow quadratically, rather than linearly.

**Reply by ggerganov** [reply]
There isn't atm. I would like to refactor the KV cache implementation to support both unified, standard and other types of cache.
